lxd/db/update.go

```diff
@@ -29,66 +30,43 @@ import (
    Only append to the updates list, never remove entries and never re-order them.
 */
 
-var dbUpdates = []dbUpdate{
-	{version: 1, run: dbUpdateFromV0},
-	{version: 2, run: dbUpdateFromV1},
-	{version: 3, run: dbUpdateFromV2},
-	{version: 4, run: dbUpdateFromV3},
-	{version: 5, run: dbUpdateFromV4},
-	{version: 6, run: dbUpdateFromV5},
-	{version: 7, run: dbUpdateFromV6},
-	{version: 8, run: dbUpdateFromV7},
-	{version: 9, run: dbUpdateFromV8},
-	{version: 10, run: dbUpdateFromV9},
-	{version: 11, run: dbUpdateFromV10},
-	{version: 12, run: dbUpdateFromV11},
-	{version: 13, run: dbUpdateFromV12},
-	{version: 14, run: dbUpdateFromV13},
-	{version: 15, run: dbUpdateFromV14},
-	{version: 16, run: dbUpdateFromV15},
-	{version: 17, run: dbUpdateFromV16},
-	{version: 18, run: dbUpdateFromV17},
-	{version: 19, run: dbUpdateFromV18},
-	{version: 20, run: dbUpdateFromV19},
-	{version: 21, run: dbUpdateFromV20},
-	{version: 22, run: dbUpdateFromV21},
-	{version: 23, run: dbUpdateFromV22},
-	{version: 24, run: dbUpdateFromV23},
-	{version: 25, run: dbUpdateFromV24},
-	{version: 26, run: dbUpdateFromV25},
-	{version: 27, run: dbUpdateFromV26},
-	{version: 28, run: dbUpdateFromV27},
-	{version: 29, run: dbUpdateFromV28},
-	{version: 30, run: dbUpdateFromV29},
-	{version: 31, run: dbUpdateFromV30},
-	{version: 32, run: dbUpdateFromV31},
-	{version: 33, run: dbUpdateFromV32},
-	{version: 34, run: dbUpdateFromV33},
-	{version: 35, run: dbUpdateFromV34},
-	{version: 36, run: dbUpdateFromV35},
-}
-
-type dbUpdate struct {
-	version int
-	run     func(previousVersion int, version int, db *sql.DB) error
-}
-
-func (u *dbUpdate) apply(currentVersion int, db *sql.DB) error {
-	// Get the current schema version
-
-	logger.Debugf("Updating DB schema from %d to %d", currentVersion, u.version)
-
-	err := u.run(currentVersion, u.version, db)
-	if err != nil {
-		return err
-	}
-
-	_, err = db.Exec("INSERT INTO schema (version, updated_at) VALUES (?, strftime(\"%s\"));", u.version)
-	if err != nil {
-		return err
-	}
-
-	return nil
+var updates = []schema.Update{
+	updateFromV0,
+	updateFromV1,
+	updateFromV2,
+	updateFromV3,
+	updateFromV4,
+	updateFromV5,
+	updateFromV6,
+	updateFromV7,
+	updateFromV8,
+	updateFromV9,
+	updateFromV10,
+	updateFromV11,
+	updateFromV12,
+	updateFromV13,
+	updateFromV14,
+	updateFromV15,
+	updateFromV16,
+	updateFromV17,
+	updateFromV18,
+	updateFromV19,
+	updateFromV20,
+	updateFromV21,
+	updateFromV22,
+	updateFromV23,
+	updateFromV24,
+	updateFromV25,
+	updateFromV26,
+	updateFromV27,
+	updateFromV28,
+	updateFromV29,
+	updateFromV30,
+	updateFromV31,
+	updateFromV32,
+	updateFromV33,
+	updateFromV34,
+	updateFromV35,
```

Comment:
Uhm, the current structure of a database update in master is:

```go
type dbUpdate struct {
	version int
	run     func(previousVersion int, version int, db *sql.DB) error
}
```

and the associated logic would make a series like this:

```go
var dbUpdates = []dbUpdate{
	{version: 1, run: foo},
	{version: 3, run: bar},
	{version: 2, run: egg},
}
```

perfectly valid: if one runs the upgrade starting from version 1, the upgrade will complete with no errors, except it will end up with version 2 stored in the schema table, and the next time a database update is run it will start from version 2 and try to re-apply version 3. One can think of other variations of this kind that end up screwing the state.

So the point is that no matter if the current code in master is more or less likely to help humans avoid mistakes, its logic is faulty, since in principle it lets those humans end up with an inconsistent state. The logic in this change does not suffer from that, as the single source of truth of the updates order is their position index in the slice.

To alleviate the situation we could perhaps go for comments:

```go
var updates = []schema.Update{
	updateFromV0, // version 1
	updateFromV1, // version 2
        // ...
}
```

Or, if you really think this might be a concern (but I'd caution against real-world evidence that it's the case), we can surely add a version field back like before. But we'd need to add some little logic that essentially ensures that there are inconsistencies between these two sources of truth (version field and position in the slice).

Alternatively, we can drop the use of a slice and have a map:

```go
// type schema.Version int
var updates = map[schema.Version]schema.Update{
	1 : updateFromV0, // version 1
	2: updateFromV1, // version 2
        // ...
}
```

and change the logic in Schema.Enforce() to make sure that there is an entry for version 1, and there are no missing versions between 1 and the highest version key in the map. It feels a bit premature defense to me, but I'll take your judgement.